optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/ConstraintCollectors.java

```diff
@@ -44,107 +46,82 @@
     // ************************************************************************
 
     public static <A> UniConstraintCollector<A, ?, Integer> count() {
-        return new DefaultUniConstraintCollector<>(
-                () -> new int[1],
-                (resultContainer, a) -> {
-                    resultContainer[0]++;
-                    return (() -> resultContainer[0]--);
-                },
-                resultContainer -> resultContainer[0]);
+        return countLong(Long::intValue);
     }
 
     public static <A> UniConstraintCollector<A, ?, Long> countLong() {
+        return countLong(Functions.identity());
+    }
+
+    private static <A, B extends Number> UniConstraintCollector<A, ?, B> countLong(Function<Long, B> countConverter) {
         return new DefaultUniConstraintCollector<>(
-                () -> new long[1],
+                LongAdder::new,
                 (resultContainer, a) -> {
-                    resultContainer[0]++;
-                    return (() -> resultContainer[0]--);
+                    resultContainer.increment();
+                    return (resultContainer::decrement);
                 },
-                resultContainer -> resultContainer[0]);
+                resultContainer -> countConverter.apply(resultContainer.longValue()));
```

Comment:
The unifying theme of this PR is that the `int[1]` result container is replaced with `LongAdder` result container.
This way, it becomes immediately obvious what the code does - in my opinion, the single-element array did not do that.